lib/pages/settings/service/admin_provider.dart

```diff
@@ -0,0 +1,123 @@
+  }
+}
+
+class AdminProvider with ChangeNotifier {
+  final FirebaseFirestore _db = FirebaseFirestore.instance;
+
+  Future<List<String>> fetchAllRequests() async {
+    try {
+      final QuerySnapshot qSnapshot = await _db
+          .collection('forms')
+          .orderBy('dateSubmitted', descending: false)
+          .get();
+      return qSnapshot.docs.map(RequestExtension.getFormId).toList();
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return null;
+    }
+  }
+
+  Future<List<String>> fetchUnprocessedRequests() async {
+    try {
+      final QuerySnapshot qSnapshot = await _db
+          .collection('forms')
+          .where('done', isEqualTo: false)
+          .orderBy('dateSubmitted', descending: false)
+          .get();
+      return qSnapshot.docs.map(RequestExtension.getFormId).toList();
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return null;
+    }
+  }
+
+  Future<Request> fetchRequestData(String requestId) async {
+    try {
+      final DocumentSnapshot docSnapshot =
+          await _db.collection('forms').doc(requestId).get();
+      return RequestExtension.fromSnap(docSnapshot);
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+      return null;
+    }
+  }
+
+  Future<User> fetchUserById(final String userId) async {
+    final snapshot =
+        await FirebaseFirestore.instance.collection('users').doc(userId).get();
+    if (snapshot.data() == null) {
+      return null;
+    }
+    final currentUser = DatabaseUser.fromSnap(snapshot);
+    return currentUser;
+  }
+
+  Future<void> acceptRequest(String formId, String userId) async {
+    try {
+      await _db.collection('users').doc(userId).update({'permissionLevel': 3});
+      await _db.collection('forms').doc(formId).update({'done': true});
+      notifyListeners();
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+    }
+  }
+
+  Future<void> denyRequest(String formId, String userId) async {
+    try {
+      await _db.collection('forms').doc(formId).update({'done': null});
+      notifyListeners();
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+    }
+  }
+
+  Future<void> revertAcceptedRequest(String formId, String userId) async {
+    try {
+      await _db.collection('users').doc(userId).update({'permissionLevel': 0});
+      await _db.collection('forms').doc(formId).update({'done': false});
+      notifyListeners();
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+    }
+  }
+
+  Future<void> revertDeniedRequest(String formId, String userId) async {
+    try {
+      await _db.collection('forms').doc(formId).update({'done': false});
+      notifyListeners();
+    } catch (e) {
+      print(e);
+      AppToast.show(S.current.errorSomethingWentWrong);
+    }
+  }
+}
```

Comment:
These should be only one method, revertRequest(requestId), which checks whether the request was approved or denied and reverts it. There should be as little logic as possible at the UI level.